### PR TITLE
Fix various crashes/issues in controller scripting API

### PR DESCRIPTION
**Pull request by HifiExperiments**
https://sentry.vircadia.dev/organizations/vircadia/issues/25/

Test plan:
- In the console, type `Controller.getAvailableInputs(9999999)`.  In master you would crash.  Now you don't crash.
- In the console, type `Controller.getActionValue(9999999)`.  In master you would crash.  Now you don't crash.

### CODE REVIEW

File: libraries/controllers/src/controllers/UserInputMapper.cpp [hunk 1]
```diff
@@ -366,7 +369,7 @@ bool UserInputMapper::triggerHapticPulse(float strength, float duration, control
     Locker locker(_lock);
     bool toReturn = false;
     for (const auto& device : _registeredDevices) {
-        toReturn = toReturn || device.second->triggerHapticPulse(strength, duration, hand);
+        toReturn = device.second->triggerHapticPulse(strength, duration, hand) || toReturn;
```
Comment (HifiExperiments) [inline libraries/controllers/src/controllers/UserInputMapper.cpp:372]: this was a logic bug we discovered a few years ago but the fix never made it in.  the old code would short-circuit after the first time toReturn was set to true, so if you had multiple devices connected, only one would buzz.  this more closely matches what the docs say this method does

File: libraries/controllers/src/controllers/UserInputMapper.cpp [hunk 2]
```diff
@@ -1237,16 +1240,42 @@ void UserInputMapper::disableMapping(const Mapping::Pointer& mapping) {
 }
 
 void UserInputMapper::setActionState(Action action, float value, bool valid) {
+    Locker locker(_lock);
```
Comment (HifiExperiments) [inline libraries/controllers/src/controllers/UserInputMapper.cpp:1243]: I don't know if these two locks are necessary but they _look_ necessary

File: libraries/controllers/src/controllers/UserInputMapper.cpp [hunk 3]
```diff
@@ -1237,16 +1240,42 @@ void UserInputMapper::disableMapping(const Mapping::Pointer& mapping) {
...
     _actionStates[toInt(action)] += delta;
     bool wasValid = _actionStatesValid[toInt(action)];
     _actionStatesValid[toInt(action)] = wasValid & valid;
 }
 
+float UserInputMapper::getActionState(Action action) const {
+    Locker locker(_lock);
+
+    int index = toInt(action);
+    if (index < _actionStates.size()) {
```
Comment (HifiExperiments) [inline libraries/controllers/src/controllers/UserInputMapper.cpp]: this is the fix to the sentry issue.  the old code didn't do bounds checking.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b.tar_d15527dd.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b.tar_d15527dd.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b_d15527dd.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b_d15527dd.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b.tar_d15527dd.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b.tar_d15527dd.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b_d15527dd.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b_d15527dd.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b_d15527dd.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b.tar_d15527dd.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b.tar_d15527dd.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b_d15527dd.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR399/ProjectAthena-Alpha-PR399-6a6e57b_d15527dd.exe

**Comment by kasenvr** [comment]
For controllers not buzzing when requested, does this apply to VR controllers as well, so only one would vibrate? If so then it should be easy to find someone to test that out, if not, then it's likely the best we can do for testing on this is to ensure that crashing no longer takes place.

Is it possible for me to link up two Xbox One controllers to my computer wirelessly and interface that way for testing?

**Comment by HifiExperiments** [comment]
@kasenvr it would affect VR controllers, but only if you had multiple left or right controllers connected, and I don't know if that's....actually possible.  theoretically it would work if you had both a rift and a vive connected, with hand controllers for both?  it's definitely not a common case, I wouldn't worry about it too much

**Comment by kasenvr** [comment]
I ask because I have no VR controllers right now. :) But if someone can test to make sure that both their controllers vibrate now then that would be perfect. I will ensure that it no longer crashes.

**Comment by SilverfishVR** [comment]
testing _PR399-6a6e57b_

Both controllers vibrate when prompted 👍 
None of the calls crash 👍 
_getAvailableInputs()_ is deprecated, I assume this intentional.
Screenshot 2020-06-18 23 00 50

**Pull request merged**